R/write.magpie.R

```diff
@@ -52,352 +52,277 @@
-      writeBin(as.integer(c(ndata*ncells*nyears,nchar(datanames_collapsed, type="bytes"))),zz,size=4)
-      if(datanames_collapsed!="") writeChar(datanames_collapsed,zz,eos=NULL)
-      writeBin(as.numeric(as.vector(x)),zz,size=4)
-      if(comment!="") writeChar(comment,zz,eos=NULL)
-      if(nchar(sets_collapsed, type="bytes")>0) writeChar(sets_collapsed,zz,eos=NULL)
-      if(!is.null(metadata)) writeBin(metadata,zz)
-      close(zz)    
-      Sys.chmod(file_path, mode)
-    } else if(file_type=="asc") {
-      coord <- magclassdata$half_deg[,c("lon","lat")]
-      if(dim(coord)[1]!=dim(x)[1]) stop("Wrong format! Only 0.5deg data can be written as ascii grid!")
-      if(any(comment!="")) warning("asc format does not support comments!")
-      for(y in 1:nyears(x)) {
-        tmp_file <- ifelse(nyears(x)>1,sub("(\\.[^\\.]*)$",paste("_",getYears(x)[y],"\\1",sep=""),file_path),file_path)
-        for(d in 1:ndata(x)) {
-          tmp2_file <- ifelse(ndata(x)>1,sub("(\\.[^\\.]*)$",paste("_",getNames(x)[d],"\\1",sep=""),tmp_file),tmp_file)
-          data <- as.data.frame(as.vector((x[,y,d])))
-          grid <- suppressWarnings(sp::SpatialPixelsDataFrame(points = coord[c("lon", "lat")], data = data))
-          sp::write.asciigrid(grid,tmp2_file)          
-        }
+
+      writeBin(as.integer(fformatVersion), zz, size = 2)
+      writeBin(as.integer(nchar(comment, type = "bytes")), zz, size = 4)
+      writeBin(as.integer(0), zz, size = 4)
+      writeBin(as.integer(nchar(setsCollapsed, type = "bytes")), zz, size = 2)
+      writeBin(as.integer(c(nyears, yearList)), zz, size = 2)
+      writeBin(as.integer(c(ncells, nchar(cellsCollapsed, type = "bytes"))), zz, size = 4)
+      writeChar(cellsCollapsed, zz, eos = NULL)
+      writeBin(as.integer(c(ndata * ncells * nyears, nchar(datanamesCollapsed, type = "bytes"))), zz, size = 4)
+      if (datanamesCollapsed != "") writeChar(datanamesCollapsed, zz, eos = NULL)
+      writeBin(as.numeric(as.vector(x)), zz, size = 4)
+      if (comment != "") writeChar(comment, zz, eos = NULL)
+      if (nchar(setsCollapsed, type = "bytes") > 0) writeChar(setsCollapsed, zz, eos = NULL)
+      close(zz)
+      Sys.chmod(filePath, mode)
+    } else if (file_type %in% c("asc", "nc", "grd", "tif")) {
+      format <- c(asc = "ascii", nc = "CDF", grd = "raster", tif = "GTiff")
+      if (ndata(x) != 1) stop("Currently no support for multiple variables for format ", format,
+                              ". Please store each variable separately.")
+      rx <- as.RasterBrick(x)
+      if (file_type == "asc") {
+        if (dim(rx)[3] != 1) stop("asc does not support multiple year layers. Please choose just one!")
+        rx <- rx[[1]]
+      }
+      varname <- getItems(x, dim = 3)
+      zunit <- ifelse(all(isYear(getYears(x))), "years", "")
+      if (is.null(varname)) varname <- "Variable"
+      raster::writeRaster(rx, filename = filePath, format = format[file_type], overwrite = TRUE,
```

Comment:
Why not `@importFrom raster writeRaster`?
This seems to happen in a couple places, maybe do a global search for `::` and check them?

Reply:
here, it was intended to not import `writeRaster` in order to achieve a weaker dependency of `magclass` on `raster` (in the DESCRIPTION it is only suggested, not imported). Doing so the package is not required in order to install `magclass`. 

This can be useful for dependencies which are only required for a small subset of the overall functionality (here, it is only required for conversions to/from raster objects and reading/writing of raster-like data formats).